dbt_subprojects/dex/macros/models/tapio_liquidity_and_volume.sql

```diff
@@ -0,0 +1,167 @@
+    pools_table = '',
+    tokens_table = 'erc20.tokens',
+    prices_table = 'prices.usd'
+) %}
+
+WITH 
+    token_swapped_events AS (
+        -- Extract data from TokenSwapped events
+        SELECT
+            block_time,
+            contract_address AS pool_address,
+            buyer AS provider,
+            swapAmount AS amount,
+            'TokenSwapped' AS event_type
+        FROM {{ source(project, blockchain, token_swapped_table) }}
+        WHERE swapAmount IS NOT NULL
+    ),
+
+    minted_events AS (
+        -- Extract data from Minted events
+        SELECT
+            block_time,
+            contract_address AS pool_address,
+            provider,
+            mintAmount AS amount,
+            'Minted' AS event_type
+        FROM {{ source(project, blockchain, minted_table) }}
+        WHERE mintAmount IS NOT NULL
+    ),
+
+    donated_events AS (
+        -- Extract data from Donated events
+        SELECT
+            block_time,
+            contract_address AS pool_address,
+            provider,
+            mintAmount AS amount,
+            'Donated' AS event_type
+        FROM {{ source(project, blockchain, donated_table) }}
+        WHERE mintAmount IS NOT NULL
+    ),
+
+    redeemed_events AS (
+        -- Extract data from Redeemed events
+        SELECT
+            block_time,
+            contract_address AS pool_address,
+            provider,
+            redeemAmount AS amount,
+            'Redeemed' AS event_type
+        FROM {{ source(project, blockchain, redeemed_table) }}
+        WHERE redeemAmount IS NOT NULL
+    ),
+
+    unified_events AS (
+        -- Combine all events into a single structure
+        SELECT * FROM token_swapped_events
+        UNION ALL
+        SELECT * FROM minted_events
+        UNION ALL
+        SELECT * FROM donated_events
+        UNION ALL
+        SELECT * FROM redeemed_events
+    ),
+
+    event_data AS (
+        -- Aggregate liquidity changes per day
+        SELECT 
+            date_trunc('day', e.block_time) AS day,
+            e.pool_address,
+            p.token0_address,
+            p.token1_address,
+            t0.symbol AS token0_symbol,
+            t1.symbol AS token1_symbol,
+            SUM(CASE 
+                WHEN e.event_type IN ('Minted', 'Donated') AND e.amount IS NOT NULL THEN CAST(e.amount AS DOUBLE)
+                ELSE 0 
+            END) AS token0_balance_raw,
+            SUM(CASE 
+                WHEN e.event_type IN ('Redeemed', 'TokenSwapped') AND e.amount IS NOT NULL THEN CAST(e.amount AS DOUBLE)
+                ELSE 0 
+            END) AS token1_balance_raw
+        FROM unified_events e
+        JOIN {{ source(blockchain, pools_table) }} p ON e.pool_address = p.address
+        JOIN {{ source(blockchain, tokens_table) }} t0 ON p.token0_address = t0.address
+        JOIN {{ source(blockchain, tokens_table) }} t1 ON p.token1_address = t1.address
+        GROUP BY 1, 2, 3, 4, 5, 6
+    )
+
+    running_balance AS (
```

Comment:
I don't think running balance is correct. Remember events give you the delta amounts added/removed to the pool. You have to `SUM() OVER()` the amounts to get balances